metabase/metabase
# feat(sdk): add visual cues in sdk around evaluation usage and feature flags

Closes https://github.com/metabase/metabase/issues/47824

### Description

Adds a visual indicator for evaluation usage (e.g. using API keys in development) and feature flags

### How to verify

Follow the test cases listed in `enterprise/frontend/src/embedding-sdk/lib/usage-problem.ts` and in the test code at `enterprise/frontend/src/embedding-sdk/components/private/SdkUsageProblem/SdkUsageProblemDisplay.unit.spec.tsx`. This is also described in [this part of the product doc](https://www.notion.so/metabase/Add-visual-cues-in-SDK-around-evaluation-usage-and-feature-flag-004be55a2ed247839cc9521e9a58f6d4?pvs=4#52b68b5b877240339a1d38bfcf0d4e0f).

For example, you can use the SDK with API keys in localhost, and the warning indicator should pop up. If you click on it, the card should open.

### Demo

Testing each test cases for visual cues

<div>
    <a href="https://www.loom.com/share/117c21d3fd954d0194855f56370b33c6">
      <p>Testing each cases in the Visual Cues PR - Private Loom</p>
    </a>
    <a href="https://www.loom.com/share/117c21d3fd954d0194855f56370b33c6">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/117c21d3fd954d0194855f56370b33c6-4da3dabb5e4544fa-full-play.gif">
    </a>
  </div>

Shows a warning when we are using API keys in development

![CleanShot 2567-09-11 at 05 17 46@2x](https://github.com/user-attachments/assets/4eea1f02-8470-4fa0-a5c9-c51470f946ad)

Shows an error when we are using API keys on production

![CleanShot 2567-09-11 at 05 48 15@2x](https://github.com/user-attachments/assets/a057efe4-5605-431f-a701-6300dc2892a4)

Renders properly in Shoppy's Pug theme

![CleanShot 2567-09-11 at 21 15 07@2x](https://github.com/user-attachments/assets/fb74246b-e554-4a50-8b6a-5efcd6fbd5cf)

Renders properly in Shoppy's Luminara theme

![CleanShot 2567-09-11 at 21 16 05@2x](https://github.com/user-attachments/assets/84d79970-e1b1-4fbd-b061-4e187701e955)

Renders properly in Shoppy's Stitch theme

![CleanShot 2567-09-11 at 21 15 52@2x](https://github.com/user-attachments/assets/77a392ad-a05e-461d-8be4-84e7ec49b07b)

### Caveat

There is a known edge case with CORS on non-localhost origins. If the user hasn't entered their license key and haven't configured their authorized origins yet, the CORS headers will be missing from the /api/session/properties endpoint, therefore all requests on non-localhost origins will fail. [We're addressing this.](https://metaboat.slack.com/archives/C063Q3F1HPF/p1726050178995979)

### Checklist
